makefile.gen

```diff
@@ -94,94 +115,104 @@ Debug: debug
 Release: release
 Asm: asm
 
-.PHONY: clean
-
 cleantmp:
 	$(RM) -f $(RES_RS)
+.PHONY: cleantmp
 
 cleandep:
 	$(RM) -f $(DEPS)
+.PHONY: cleandep
 
 cleanlst:
 	$(RM) -f $(LSTS)
+.PHONY: cleanlst
 
 cleanres: cleantmp
 	$(RM) -f $(RES_H) $(RES_DEP) $(RES_DEPS)
+.PHONY: cleanres
 
 cleanobj:
-	$(RM) -f $(OBJS) out/sega.o out/rom_head.bin out/rom_head.o out/rom.out
+	$(RM) -f $(OBJS) $(OUT)/sega.o $(OUT)/rom_head.bin $(OUT)/rom_head.o $(OUT)/rom.out
+.PHONY: cleanobj
 
 clean: cleanobj cleanres cleanlst cleandep
-	$(RM) -f out.lst out/cmd_ out/symbol.txt out/rom.nm out/rom.wch out/rom.bin
+	$(RM) -f out.lst $(OUT)/cmd_ $(OUT)/symbol.txt $(OUT)/rom.nm $(OUT)/rom.wch $(OUT)/rom.bin
+.PHONY: clean
 
 cleanrelease: clean
+.PHONY: cleanrelease
 
 cleandebug: clean
+.PHONY: cleandebug
 
 cleanasm: cleanlst
+.PHONY: cleanasm
 
 cleandefault: clean
 cleanDefault: clean
+.PHONY: cleandefault cleanDefault
 
 cleanRelease: cleanrelease
 cleanDebug: cleandebug
 cleanAsm: cleanasm
-
-pre-build:
-	$(MKDIR) -p $(SRC)/boot
-	$(MKDIR) -p out
+.PHONY: cleanRelease cleanDebug cleanAsm
 
 
-out/rom.bin: out/rom.out
-	$(OBJCPY) -O binary out/rom.out out/rom.bin
-	$(SIZEBND) out/rom.bin -sizealign 131072 -checksum
+$(OUT)/rom.bin: $(OUT)/rom.out
+	$(OBJCPY) -O binary $(OUT)/rom.out $(OUT)/rom.bin
+	$(SIZEBND) $(OUT)/rom.bin -sizealign 131072 -checksum
 
-out/symbol.txt: out/rom.out
-	$(NM) $(LTO_PLUGIN) -n out/rom.out > out/symbol.txt
+$(OUT)/symbol.txt: $(OUT)/rom.out
+	$(NM) $(LTO_PLUGIN) -n $(OUT)/rom.out > $(OUT)/symbol.txt
 
-out/rom.out: out/sega.o out/cmd_ $(LIBMD)
-	$(CC) -m68000 -B$(BIN) -n -T $(GDK)/md.ld -nostdlib out/sega.o @out/cmd_ $(LIBMD) $(LIBGCC) -o out/rom.out -Wl,--gc-sections -flto
-	$(RM) out/cmd_
+$(OUT)/rom.out: $(OUT)/sega.o $(OUT)/cmd_ $(LIBMD)
+	$(MKDIR) -p $(dir $@)
+	$(CC) -m68000 -B$(BIN) -n -T $(GDK)/md.ld -nostdlib $(OUT)/sega.o @$(OUT)/cmd_ $(LIBMD) $(LIBGCC) -o $(OUT)/rom.out -Wl,--gc-sections -flto
+	$(RM) $(OUT)/cmd_
 
-out/cmd_: $(OBJS)
-	$(ECHO) "$(OBJS)" > out/cmd_
+$(OUT)/cmd_: $(OBJS)
+	$(MKDIR) -p $(dir $@)
+	$(ECHO) "$(OBJS)" > $(OUT)/cmd_
 
-out/sega.o: $(SRC)/boot/sega.s out/rom_head.bin
-	$(CC) -x assembler-with-cpp -Wa,--register-prefix-optional,--bitwise-or $(AFLAGS) -c $(SRC)/boot/sega.s -o $@
+$(OUT)/sega.o: $(OUT)/boot/sega.s $(OUT)/rom_head.bin
+	$(CC) -x assembler-with-cpp -Wa,--register-prefix-optional,--bitwise-or $(AFLAGS) -c $(OUT)/boot/sega.s -o $@
 
-out/rom_head.bin: out/rom_head.o
+$(OUT)/rom_head.bin: $(OUT)/rom_head.o
 	$(OBJCPY) -O binary $< $@
 
-out/rom_head.o: $(SRC)/boot/rom_head.c
+$(OUT)/rom_head.o: $(SRC)/boot/rom_head.c
+	$(MKDIR) -p $(dir $@)
 	$(CC) $(DEFAULT_FLAGS) -c $< -o $@
 
-$(SRC)/boot/sega.s: $(SRC_LIB)/boot/sega.s
+$(OUT)/boot/sega.s: $(SRC_LIB)/boot/sega.s

```

Comment:
Why did you moved boot/sega.s to _OUT_ folder ? The file make be customized (interrupt callback, bootstrap..) so i think it's important to keep it in _SRC_ folder

Reply:
I inherited most of these changes from the original PR.  I just tried to clean it up for review and fix the workflows.

I'm guessing the original author thought that generated files should go to OUT as a rule, and didn't realize that it could be customized.  I would have made the same mistake.

I'll revert this.

Reply:
I have run into one problem with these generated sources, though, which I will try to resolve.  If I rebuild SGDK, my existing src/boot/rom_head.c gets overwritten when I rebuild my ROM because the one in SRC_LIB is newer.  I only noticed because I had already customized my rom_head.c file.

Reply:
Reverted sega.s moving to OUT.  Fixed overwritten boot files.